backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/window/descriptor/decorator/ExternallyReferencedDocumentDecorator.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * #%L
+ * de.metas.ui.web.base
+ * %%
+ * Copyright (C) 2022 metas GmbH
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program. If not, see
+ * <http://www.gnu.org/licenses/gpl-2.0.html>.
+ * #L%
+ */
+
+package de.metas.ui.web.window.descriptor.decorator;
+
+import de.metas.externalreference.ExternalReference;
+import de.metas.externalreference.ExternalReferenceRepository;
+import de.metas.i18n.AdMessageKey;
+import de.metas.i18n.IMsgBL;
+import de.metas.i18n.ITranslatableString;
+import de.metas.ui.web.window.model.Document;
+import de.metas.util.Services;
+import lombok.NonNull;
+import org.adempiere.util.lang.impl.TableRecordReference;
+import org.springframework.stereotype.Service;
+
+import java.util.Optional;
+
+@Service
+public class ExternallyReferencedDocumentDecorator implements IDocumentDecorator
+{
+	private static final AdMessageKey EXTERNAL_REFERENCE_READ_ONLY_IN_METASFRESH_ERROR = AdMessageKey.of("externalReferenceReadOnlyInMetasfresh");
+
+	private final IMsgBL msgBL = Services.get(IMsgBL.class);
+
+	private final ExternalReferenceRepository externalReferenceRepository;
+
+	public ExternallyReferencedDocumentDecorator(@NonNull final ExternalReferenceRepository externalReferenceRepository)
+	{
+		this.externalReferenceRepository = externalReferenceRepository;
+	}
+
+	@Override
+	public boolean isReadOnly(final Document document)
+	{
+		final TableRecordReference recordReference = document.getTableRecordReference().orElse(null);
+
+		if (recordReference == null)
+		{
+			return false;
+		}
+
+		return externalReferenceRepository.getExternalReferencesByTableRecordReference(recordReference)
+				.stream()
+				.anyMatch(ExternalReference::getIsReadOnlyInMetasfreshOrFalse);
```

Comment:
IMHO this shall be super efficient, 
so instead of `get list` then `stream` then `anyMatch`, 
I would suggest checking if matching directly at the database level, no need to load it as a list...

Reply:
so we shall introduce a method, something like 
ExternalReferenceRepository.isReadonly(TableRecordReference recordReference).
That would be first step of optimization.

Now, keep in mind
* this method will be called on every record the webui will check
* this feature of readonly external references will be used my very few customers and when used it will be used only very few entities (e.g. only BPartners will be made readonly because are maintained externally)

Having this in mind, we can minime the performance impact of other customers and even on all other records tables as follows:

In ExternalReferenceRepository maintain a cached Set of TableNames (or AD_Table_IDs, or both?) which have records that have external references with IsReadOnlyInMetasfresh=true.
Whenever ExternalReferenceRepository.isReadonly is called, first check our cached set. 
If the tablename is not in that set, we know for sure that our record is NOT readonly.
If the tablename is in that set, we have to check in database.